flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManager.java

```diff
@@ -0,0 +1,287 @@
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An active implementation of {@link ResourceManager}.
+ *
+ * <p>This resource manager actively requests and releases resources from/to the external resource management frameworks.
+ * With different {@link ResourceManagerDriver} provided, this resource manager can work with various frameworks.
+ */
+public class ActiveResourceManager<WorkerType extends ResourceIDRetrievable>
+		extends ResourceManager<WorkerType> implements ResourceEventHandler<WorkerType> {
+
+	protected final Configuration flinkConfig;
+
+	private final ResourceManagerDriver<WorkerType> resourceManagerDriver;
+
+	/** All workers maintained by {@link ActiveResourceManager}. */
+	private final Map<ResourceID, WorkerType> workerNodeMap;
+
+	/** Number of requested and not registered workers per worker resource spec. */
+	private final PendingWorkerCounter pendingWorkerCounter;
+
+	/** Identifiers and worker resource spec of requested not registered workers. */
+	private final Map<ResourceID, WorkerResourceSpec> currentAttemptUnregisteredWorkers;
+
+	public ActiveResourceManager(
+			ResourceManagerDriver<WorkerType> resourceManagerDriver,
+			Configuration flinkConfig,
+			RpcService rpcService,
+			ResourceID resourceId,
+			HighAvailabilityServices highAvailabilityServices,
+			HeartbeatServices heartbeatServices,
+			SlotManager slotManager,
+			ResourceManagerPartitionTrackerFactory clusterPartitionTrackerFactory,
+			JobLeaderIdService jobLeaderIdService,
+			ClusterInformation clusterInformation,
+			FatalErrorHandler fatalErrorHandler,
+			ResourceManagerMetricGroup resourceManagerMetricGroup) {
+		super(
+				rpcService,
+				resourceId,
+				highAvailabilityServices,
+				heartbeatServices,
+				slotManager,
+				clusterPartitionTrackerFactory,
+				jobLeaderIdService,
+				clusterInformation,
+				fatalErrorHandler,
+				resourceManagerMetricGroup,
+				AkkaUtils.getTimeoutAsTime(Preconditions.checkNotNull(flinkConfig)));
+
+		this.flinkConfig = flinkConfig;
+		this.resourceManagerDriver = resourceManagerDriver;
+		this.workerNodeMap = new HashMap<>();
+		this.pendingWorkerCounter = new PendingWorkerCounter();
+		this.currentAttemptUnregisteredWorkers = new HashMap<>();
+	}
+
+	// ------------------------------------------------------------------------
+	//  ResourceManager
+	// ------------------------------------------------------------------------
+
+	@Override
+	protected void initialize() throws ResourceManagerException {
+		try {
+			resourceManagerDriver.initialize(
+					this,
+					command -> getMainThreadExecutor().execute(command)); // always execute on the current main thread executor.
```

Comment:
Just as a note: If the `ResourceManagerDriver` tries to run something on this executor while the `ResourceManager` has no leadership, then it will simply be ignored and not executed. Is this ok with the current behaviour of the `ResourceManagerDriver`?

Reply:
Yes, double checked.